File: src/network/libqhstspolicy.zig
const C = @import("qt6c");
const qurl_enums = @import("../libqurl.zig").enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qhstspolicy.html
pub const qhstspolicy = struct {
    /// New constructs a new QHstsPolicy object.
    ///
    ///
    pub fn New() ?*C.QHstsPolicy {
        return C.QHstsPolicy_new();
    }

    /// New2 constructs a new QHstsPolicy object.
    ///
    /// ``` expiry: ?*C.QDateTime, flags: i32, host: []const u8 ```
    pub fn New2(expiry: ?*anyopaque, flags: i64, host: []const u8) ?*C.QHstsPolicy {
        const host_str = C.struct_libqt_string{
            .len = host.len,
            .data = @constCast(host.ptr),
        };

        return C.QHstsPolicy_new2(@ptrCast(expiry), @intCast(flags), host_str);
    }

    /// New3 constructs a new QHstsPolicy object.
    ///
    /// ``` rhs: ?*C.QHstsPolicy ```
    pub fn New3(rhs: ?*anyopaque) ?*C.QHstsPolicy {
        return C.QHstsPolicy_new3(@ptrCast(rhs));
    }

    /// New4 constructs a new QHstsPolicy object.
    ///
    /// ``` expiry: ?*C.QDateTime, flags: i32, host: []const u8, mode: qurl_enums.ParsingMode ```
    pub fn New4(expiry: ?*anyopaque, flags: i64, host: []const u8, mode: i64) ?*C.QHstsPolicy {
        const host_str = C.struct_libqt_string{
            .len = host.len,
            .data = @constCast(host.ptr),
        };

        return C.QHstsPolicy_new4(@ptrCast(expiry), @intCast(flags), host_str, @intCast(mode));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#operator=)
    ///
    /// ``` self: ?*C.QHstsPolicy, rhs: ?*C.QHstsPolicy ```
    pub fn OperatorAssign(self: ?*anyopaque, rhs: ?*anyopaque) void {
        C.QHstsPolicy_OperatorAssign(@ptrCast(self), @ptrCast(rhs));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#swap)
    ///
    /// ``` self: ?*C.QHstsPolicy, other: ?*C.QHstsPolicy ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QHstsPolicy_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#setHost)
    ///
    /// ``` self: ?*C.QHstsPolicy, host: []const u8 ```
    pub fn SetHost(self: ?*anyopaque, host: []const u8) void {
        const host_str = C.struct_libqt_string{
            .len = host.len,
            .data = @constCast(host.ptr),
        };
        C.QHstsPolicy_SetHost(@ptrCast(self), host_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#host)
    ///
    /// ``` self: ?*C.QHstsPolicy, allocator: std.mem.Allocator ```
    pub fn Host(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QHstsPolicy_Host(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#setExpiry)
    ///
    /// ``` self: ?*C.QHstsPolicy, expiry: ?*C.QDateTime ```
    pub fn SetExpiry(self: ?*anyopaque, expiry: ?*anyopaque) void {
        C.QHstsPolicy_SetExpiry(@ptrCast(self), @ptrCast(expiry));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#expiry)
    ///
    /// ``` self: ?*C.QHstsPolicy ```
    pub fn Expiry(self: ?*anyopaque) ?*C.QDateTime {
        return C.QHstsPolicy_Expiry(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#setIncludesSubDomains)
    ///
    /// ``` self: ?*C.QHstsPolicy, include: bool ```
    pub fn SetIncludesSubDomains(self: ?*anyopaque, include: bool) void {
        C.QHstsPolicy_SetIncludesSubDomains(@ptrCast(self), include);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#includesSubDomains)
    ///
    /// ``` self: ?*C.QHstsPolicy ```
    pub fn IncludesSubDomains(self: ?*anyopaque) bool {
        return C.QHstsPolicy_IncludesSubDomains(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#isExpired)
    ///
    /// ``` self: ?*C.QHstsPolicy ```
    pub fn IsExpired(self: ?*anyopaque) bool {
        return C.QHstsPolicy_IsExpired(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#setHost)
    ///
    /// ``` self: ?*C.QHstsPolicy, host: []const u8, mode: qurl_enums.ParsingMode ```
    pub fn SetHost2(self: ?*anyopaque, host: []const u8, mode: i64) void {
        const host_str = C.struct_libqt_string{
            .len = host.len,
            .data = @constCast(host.ptr),
        };
        C.QHstsPolicy_SetHost2(@ptrCast(self), host_str, @intCast(mode));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qhstspolicy.html#host)
    ///
    /// ``` self: ?*C.QHstsPolicy, options: u32, allocator: std.mem.Allocator ```
    pub fn Host1(self: ?*anyopaque, options: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QHstsPolicy_Host1(@ptrCast(self), @intCast(options));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QHstsPolicy ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QHstsPolicy_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qhstspolicy.html#types
pub const enums = struct {
    pub const PolicyFlag = enum {
        pub const IncludeSubDomains: i32 = 1;
    };
};

File: src/network/libqhstspolicy.cpp
#include <QDateTime>
#include <QHstsPolicy>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qhstspolicy.h>
#include "libqhstspolicy.h"
#include "libqhstspolicy.hxx"

QHstsPolicy* QHstsPolicy_new() {
    return new QHstsPolicy();
}

QHstsPolicy* QHstsPolicy_new2(const QDateTime* expiry, int flags, const libqt_string host) {
    QString host_QString = QString::fromUtf8(host.data, host.len);
    return new QHstsPolicy(*expiry, static_cast<QHstsPolicy::PolicyFlags>(flags), host_QString);
}

QHstsPolicy* QHstsPolicy_new3(const QHstsPolicy* rhs) {
    return new QHstsPolicy(*rhs);
}

QHstsPolicy* QHstsPolicy_new4(const QDateTime* expiry, int flags, const libqt_string host, int mode) {
    QString host_QString = QString::fromUtf8(host.data, host.len);
    return new QHstsPolicy(*expiry, static_cast<QHstsPolicy::PolicyFlags>(flags), host_QString, static_cast<QUrl::ParsingMode>(mode));
}

void QHstsPolicy_OperatorAssign(QHstsPolicy* self, const QHstsPolicy* rhs) {
    self->operator=(*rhs);
}

void QHstsPolicy_Swap(QHstsPolicy* self, QHstsPolicy* other) {
    self->swap(*other);
}

void QHstsPolicy_SetHost(QHstsPolicy* self, const libqt_string host) {
    QString host_QString = QString::fromUtf8(host.data, host.len);
    self->setHost(host_QString);
}

libqt_string QHstsPolicy_Host(const QHstsPolicy* self) {
    QString _ret = self->host();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QHstsPolicy_SetExpiry(QHstsPolicy* self, const QDateTime* expiry) {
    self->setExpiry(*expiry);
}

QDateTime* QHstsPolicy_Expiry(const QHstsPolicy* self) {
    return new QDateTime(self->expiry());
}

void QHstsPolicy_SetIncludesSubDomains(QHstsPolicy* self, bool include) {
    self->setIncludesSubDomains(include);
}

bool QHstsPolicy_IncludesSubDomains(const QHstsPolicy* self) {
    return self->includesSubDomains();
}

bool QHstsPolicy_IsExpired(const QHstsPolicy* self) {
    return self->isExpired();
}

void QHstsPolicy_SetHost2(QHstsPolicy* self, const libqt_string host, int mode) {
    QString host_QString = QString::fromUtf8(host.data, host.len);
    self->setHost(host_QString, static_cast<QUrl::ParsingMode>(mode));
}

libqt_string QHstsPolicy_Host1(const QHstsPolicy* self, int options) {
    QString _ret = self->host(static_cast<QUrl::ComponentFormattingOptions>(options));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

void QHstsPolicy_Delete(QHstsPolicy* self) {
    delete self;
}
